imap/ical_support.c

```diff
@@ -3260,4 +3273,131 @@ EXPORTED void icalcomponent_set_jmapid(icalcomponent *comp, const char *id)
 }
 #endif
 
+#ifdef HAVE_RECUR_BY_REF
+EXPORTED icalrecurrencetype_t *icalvalue_get_recurrence(const icalvalue *val)
+{
+    icalrecurrencetype_t *rt = icalvalue_get_recur(val);
+
+    icalrecurrencetype_ref(rt);
+
+    return rt;
+}
+
+#else /* !HAVE_RECUR_BY_REF */
+EXPORTED icalrecurrencetype_t *icalrecurrencetype_new(void)
+{
+    icalrecurrencetype_t *rt =
+        icalmemory_new_buffer(sizeof(icalrecurrencetype_t));
+
+    if (!rt) return NULL;
+
+    icalrecurrencetype_clear(rt);
+
+    return rt;
+}
+
+EXPORTED void icalrecurrencetype_unref(icalrecurrencetype_t *rt)
+{
+    icalerror_check_arg_rv((rt != NULL), "rt");
+
+    if (rt->rscale) icalmemory_free_buffer(rt->rscale);
+
+    icalmemory_free_buffer(rt);
+}
+
+EXPORTED icalrecurrencetype_t *icalrecurrencetype_clone(icalrecurrencetype_t *rt)
+{
+    icalrecurrencetype_t *res;
+
+    icalerror_check_arg_rz((rt != NULL), "rt");
+
+    res = icalrecurrencetype_new();
+    if (!res) {
+        return NULL;
+    }
+
+    memcpy(res, rt, sizeof(*res));
+
+    if (res->rscale) {
+        res->rscale = icalmemory_strdup(res->rscale);
+        if (!res->rscale) {
+            icalrecurrencetype_unref(res);
+            return NULL;
+        }
+    }
+
+    return res;
+}
+
+EXPORTED icalrecurrencetype_t *icalrecurrencetype_new_from_string(const char *str)
+{
+    icalrecurrencetype_t rt = icalrecurrencetype_from_string(str);
+    icalrecurrencetype_t *res = icalrecurrencetype_clone(&rt);
+
+    if (rt.rscale) icalmemory_free_buffer(rt.rscale);
+
+    return res;
+}
+
+EXPORTED icalrecurrencetype_t *icalvalue_get_recurrence(const icalvalue *val)
+{
+    icalrecurrencetype_t rt = icalvalue_get_recur(val);
+
+    return icalrecurrencetype_clone(&rt);
+}
+
+EXPORTED short *icalrecur_byrule_data(icalrecurrencetype_t *rt,
+                                      icalrecurrencetype_byrule rule)
+{
+    switch (rule) {
+    case ICAL_BY_MONTH:     return rt->by_month;
+    case ICAL_BY_WEEK_NO:   return rt->by_week_no;
+    case ICAL_BY_YEAR_DAY:  return rt->by_year_day;
+    case ICAL_BY_MONTH_DAY: return rt->by_month_day;
+    case ICAL_BY_DAY:       return rt->by_day;
+    case ICAL_BY_HOUR:      return rt->by_hour;
+    case ICAL_BY_MINUTE:    return rt->by_minute;
+    case ICAL_BY_SECOND:    return rt->by_second;
+    case ICAL_BY_SET_POS:   return rt->by_set_pos;
+    default:                return NULL;
+    }
+}
+
+EXPORTED short icalrecur_byrule_size(icalrecurrencetype_t *rt,
+                                     icalrecurrencetype_byrule rule)
+{
+    switch (rule) {
+    case ICAL_BY_MONTH:
+        if (rt->by_month[0] == ICAL_RECURRENCE_ARRAY_MAX) return 0;
+        else return ICAL_BY_MONTH_SIZE;
```

Comment:
Shouldn't this check for the position of the first occurrence of ICAL_RECURRENCE_ARRAY_MAX in the by_xxx array and return that, falling back to ICAL_BY_XXX_SIZE only if none could be found in the first ICAL_BY_XXX_SIZE positions?

This is what's [documented for version 3.0](https://github.com/libical/libical/blob/e8b159db8dcf25371689e4cabb7996aa86f600a9/src/libical/icalrecur.h#L168-L170): 
```
     * Each of the lists is terminated with the value
     * ICAL_RECURRENCE_ARRAY_MAX unless the list is full.
     */
```

Instead, the version 4.0 implementation does not make such guarantee.

Reply:
So, good point.  The existing code that uses this call already checks for ICAL_RECURRENCE_ARRAY_MAX, but I guess I should just move that here and then we won't be checking for that value for the new dynamic arrays.